source/slang/slang-ast-base.h

```diff
@@ -116,6 +124,177 @@ class SyntaxNodeBase : public NodeBase
     SourceLoc loc;
 };
 
+enum class ValNodeOperandKind
+{
+    ConstantValue,
+    ValNode,
+    ASTNode,
```

Comment:
It's interesting, because we really do have the makings of a strong distinction between the *syntax* nodes (inherently have source locations, etc.) and the `Val`s (which don't).

Since syntax nodes represent actual fragments of source code (or at least can), they don't really *want* deduplication (or we could have them hold a deduplicated "body" along with a single starting source location...).

In contrast, it should always be reasonable to deduplicate `Val`s, even if we don't end up relying on the deduplication for identity testing.